src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -1291,18 +1328,63 @@ private Task WriteInternalAsync(byte[] array, int offset, int numBytes, Cancella
                 _readLen = 0;
             }
 
-            int n = _bufferSize - _writePos;
-            if (numBytes <= n)
+            // There are a few different cases to handle when performing the write operation
+            // surrounding the internal buffer. Buffer flush operations can be issued asynchronously
+            // so the state of any async operation that touches the internal buffer must be tracked
+            // to ensure there are no data races at play.
+            //
+            //   1. No active async buffer op and numBytes fits in existing buffer:
+            //       Simply perform a memory copy operation and return synchronously.
+            //
+            //   2. Active async buffer op and numBytes <= _bufferSize
+            //       Since there's an active buffer operation, a new buffer needs to be allocated
+            //       to write to in order to ensure the existing buffer isn't modified mid-write.
+            //       The incoming write can then be copied into the new buffer.
+            //
+            //   3. No active async buffer and numBytes too large for buffer
+            //       If there's buffered data, issue a flush operation and then directly issue the
+            //       incoming write since it can't be buffered.
+            //
+            //   4. Active async buffer op and numBytes too large for buffer
+            //       If there's buffered data, attach it to the existing async operation using Task.WhenAll
+            //       and then directly issue the incoming write.
+            //
+            //       Note: since _writePos is reset synchronously when calling FlushWriteAsync, it can be
+            //       assumed that if _writePos > 0 and there's an active async flush operation, the buffer
+            //       was reallocated since that operation was initiated.
+
+            // Case 1 - no active buffer op & fits in remaining size
+            int remainingBuffer = _bufferSize - _writePos;
+            if(!ActiveBufferOperation() && numBytes <= remainingBuffer)
             {
-                if (_writePos == 0) _buffer = new byte[_bufferSize];
+                if (_buffer == null) _buffer = new byte[_bufferSize];
+
                 Buffer.BlockCopy(array, offset, _buffer, _writePos, numBytes);
                 _writePos += numBytes;
-                
                 return Task.CompletedTask;
             }
 
+            // Case 2 - active buffer op & fits in new buffer
+            else if (ActiveBufferOperation() && numBytes <= _bufferSize)
```

Comment:
Consider getting rid of the ```else``` here for case 2: it's not needed because case 1 ends with a return, and case 3/4 don't have an else, so this (or those) are inconsistent.